### PR TITLE
Implement delta indexes

### PR DESCRIPTION
**Pull request by calmh**
Not done yet, but works so far... Some background in http://docs.syncthing.net/draft/localver.html

### CODE REVIEW

**Comment by calmh** [comment]
@st-jenkins run integration tests

File: internal/model/model.go [hunk 1]
```diff
@@ -704,6 +705,24 @@ func (m *Model) ClusterConfig(deviceID protocol.DeviceID, cm protocol.ClusterCon
 	if changed {
 		m.cfg.Save()
 	}
+
+	m.fmut.RLock()
+	for _, folder := range cm.Folders {
+		fs := m.folderFiles[folder.ID]
+		var initialLocalVer int64
+		for _, device := range folder.Devices {
+			if bytes.Compare(device.ID, m.id[:]) == 0 {
```
Comment (AudriusButkevicius) [inline internal/model/model.go]: bytes.Equal?

**Comment by calmh** [comment]
Note to self:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x28 pc=0x602a89]

goroutine 54 [running]:
github.com/syncthing/syncthing/internal/model.sendIndexes(0x0, 0x0, 0xc208122540, 0x7, 0xc20810a200, 0xc2080705f0, 0x0)
	/go/src/github.com/syncthing/syncthing/internal/model/model.go:1046 +0x59
created by github.com/syncthing/syncthing/internal/model.(*Model).ClusterConfig
	/go/src/github.com/syncthing/syncthing/internal/model/model.go:722 +0x142f
```

connection may receive ClusterConfig before it's been added to the Model because raisins

**Comment by AudriusButkevicius** [comment]
I had the exact same problem working on temp indexes.

**Comment by calmh** [comment]
Yeah I've been thinking about it and honestly the protocol package design isn't awesome in this aspect. I think at minimum it should start it's goroutines from a `ServeBackground()` or similar instead of in the constructor - then we can add it properly first and start it after. And I'm a bit suspicious towards its use of `go` on some callbacks - that feels like it should be handled by the model (nowadays) if stuff should run in the background. I'll try to clean that up and see how it looks.

**Comment by calmh** [comment]
This could stand another pair of eyes again as I think it's fairly close to done. Benefits from whitespace insensitive diff (`&w=1` on Github or `git show -w`...).

File: internal/model/model.go [hunk 2]
```diff
@@ -637,73 +638,115 @@ func (m *Model) ClusterConfig(deviceID protocol.DeviceID, cm protocol.ClusterCon
 		}
 	}
 
-	if m.cfg.Devices()[deviceID].Introducer {
```
Comment (calmh) [inline internal/model/model.go]: This stuff is just moved unchanged into a method because model.ClusterConfig() was getting large.

**Comment by AudriusButkevicius** [comment]
Ill go through this over the weekend again.

**Comment by AudriusButkevicius** [comment]
So sadly I understand to little about this to make a good judgement. I to this day haven't got the grips of these local versions. It doesn't seem check the local verison received part of cm?

**Comment by Zillode** [comment]
Request for clarifications:

1. *If the received maximum local version number is lower than the maximum local version number in our index, the other device is missing some of our index information. We send an incremental index (“Index Update” messages only) containing all index entries with higher local version numbers than the received maximum.*
This means Index Update exchange needs to be sorted by local version number. However, `sendIndexTo` seems to be missing this constraint.

2. I'm unsure under what conditions the statement ```file.LocalVersion == 0``` is true.

3. I'm unsure under what conditions the statement ```if s.minLocalVersion[deviceID] == 0``` is true.

4. Where does ```initialLocalVer = device.MaxLocalVersion``` for ```resetting to zero``` come from? I'd like to assess whether this could be subject to a race condition.

It would be nice to run ```test/sync_test.go``` with additional random restarts during the sync process to further validate the correctness of the implementation.

**Comment by calmh** [comment]
> This means Index Update exchange needs to be sorted by local version number.

Hmm, yes, in case a connection is dropped during the initial index update... That's a bit annoying indeed.

> I'm unsure under what conditions the statement file.LocalVersion == 0 is true.

Basically when we get a new (updated) FileInfo object from the scanner or after pulling a file. It should be taken to mean "assign a new local version"

> I'm unsure under what conditions the statement if s.minLocalVersion[deviceID] == 0 is true.

The first time we look at it, when reading the database at initialization. :)

> Where does initialLocalVer = device.MaxLocalVersion for resetting to zero come from? I'd like to assess whether this could be subject to a race condition.

What line of code are you referring to? There's a check for whether the other side is claiming to have index entries of ours with higher version than we have, which we know is impossible. So if we see that, we consider their index suspicious and do the full exchange (as if they had announced local version = 0).

**Comment by calmh** [comment]
> It would be nice to run test/sync_test.go with additional random restarts during the sync process to further validate the correctness of the implementation.

I've been stressing it with the reconnect_test.go stuff, setting a fairly low bandwidth limit to get lots of reconnects on either side. It wouldn't hurt to extend that to performing some changes now and then too I guess.

I'll probably need to refactor the reconnect_test a bit as part of this anyway, as it currently only works properly on the build server when built with -race, otherwise the sync goes too fast for the reconnects to actually happen...

**Comment by calmh** [comment]
> It doesn't seem check the local verison received part of cm?

Sure it does. :) `startIndexSenders` as called from `ClusterConfig` digs out what the other party claims to have for us and starts the index sender with that as the current local version number.

**Comment by calmh** [comment]
> This means Index Update exchange needs to be sorted by local version number.

I'm thinking of maintaining a separate (in database) `{folder ID, local version number} -> file name` table. This requires updating when we do updates, but it then trivially answers "which files are updated since $some-local-version" (just seek to that version number and start iterating) which would be useful for the index senders. Also makes sure the sent updates are sorted correctly.

**Comment by Zillode** [comment]
Good idea, I was actually reviving the sqlite3 branch to measure/improve the performance of `need()`

**Comment by calmh** [comment]
Adding an `indexSorter` to handle this. The idea being this acts like a sponge during index sending - accumulating all the FileInfos, then sorting and batching them. For small indexes this stays in RAM, then at some point it can spill to a temporary DB on disk for the sorting.

**Comment by calmh** [comment]
@st-jenkins run integration tests please

File: lib/model/indexsorter.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+			Name         string // max:8192
+			Flags        uint32
+			Modified     int64
+			Version      Vector
+			LocalVersion int64
+			CachedSize   int64       // noencode (cache only)
+			Blocks       []BlockInfo // max:1000000
+		}
+	*/
+	return 8 + len(f.Name) + 4 + 8 + 8 + 16*len(f.Version) + 8 + 8 + 8 + 8 + 40*len(f.Blocks)
```
Comment (AudriusButkevicius) [inline lib/model/indexsorter.go]: I suggest you use unsafe.SizeOf() for the struct, and check that it matches the expected value, and panic if it doesn't, then add name, blocks and versions separately.

File: lib/model/indexsorter.go [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+		db:     db,
+		dbPath: path,
+	}, nil
+}
+
+func (s *leveldbIndexSorter) Enqueue(items ...protocol.FileInfo) {
+	// Marshal each item, keyed by it's LocalVersion.
+	for _, f := range items {
+		var key [8]byte
+		binary.BigEndian.PutUint64(key[:], uint64(f.LocalVersion))
```
Comment (AudriusButkevicius) [inline lib/model/indexsorter.go]: LocalVersion is unique per file, right?

File: lib/model/indexsorter.go (hunk 3 above)
Comment (calmh) [inline lib/model/indexsorter.go]: Thanks, I'd forgotten about unsafe.SizeOf!

File: lib/model/indexsorter.go (hunk 4 above)
Comment (calmh) [inline lib/model/indexsorter.go]: As of [this](https://github.com/syncthing/syncthing/pull/2020/files#diff-69b15624219bb5333982e6cf1df4ef78R32) it's globally unique, however there may be an issue here if the device's clock is set backwards and Syncthing restarted...

**Comment by Nutomic** [comment]
Just curious, is this still coming? :)

**Comment by calmh** [comment]
At some point probably, it's just a lot more work than it initially seemed to be. :)

**Comment by calmh** [comment]
To be revived when time permits

**Pull request closed without merging**